libvast/src/experimental_table_slice_builder.cpp

```diff
@@ -502,6 +502,50 @@ class subnet_column_builder
   std::shared_ptr<arrow::StructBuilder> subnet_builder_;
 };
 
+/// Consumes arrow arrays and builds up the appropriate array based on the
+/// provided data type.
+/// @param array_iterator An iterator to consume column arrays.
+/// @param arrow_type the data type we're building the array for.
+std::shared_ptr<arrow::Array>
+make_arrow_array(arrow::ArrayVector::const_iterator& array_iterator,
+                 std::shared_ptr<arrow::DataType> arrow_type) {
+  auto f = detail::overload{
+    // TODO: catch-all is bad, but implementing 15 code paths for standard
+    // arrays ain't any better
+    [&](const std::shared_ptr<arrow::DataType>&)
+      -> std::shared_ptr<arrow::Array> {
+      return *array_iterator++;
+    },
+    [&](const std::shared_ptr<arrow::StructType>& st)
+      -> std::shared_ptr<arrow::Array> {
+      arrow::ArrayVector children{};
+      std::vector<std::string> field_names{};
+      children.reserve(st->num_fields());
+      field_names.reserve(st->num_fields());
+      for (const auto& field : st->fields()) {
+        children.push_back(make_arrow_array(array_iterator, field->type()));
+        field_names.push_back(field->name());
+      }
+      auto res = arrow::StructArray::Make(children, field_names);
+      return res.ValueOrDie();
+    }};
+  return caf::visit(f, arrow_type);
+}
+
+auto make_record_batch(const arrow::ArrayVector& columns, size_t rows,
+                       const std::shared_ptr<arrow::Schema>& schema) {
+  auto it = columns.begin();
+  auto output_columns = arrow::ArrayVector{};
+  output_columns.reserve(schema->num_fields());
+  for (const auto& field : schema->fields())
+    output_columns.push_back(make_arrow_array(it, field->type()));
+  // TODO: try this out, can we re-use the logic and just call into
+  // make_arrow_array directly? arrow::RecordBatch::FromStructArray(const
+  // std::shared_ptr<Array> &array)
```

Comment:
It would allow us to get rid of the distinction between schema (top-level struct) and normal / nested structs. However, on revisiting I don't see a way to attach meta data information at top level (type names and attributes), because the top-level entity to attach that information on seems to be missing.